File: src/service/product.ts
import { invoke } from "@tauri-apps/api/core";
import { getErrorMessage } from "@/lib/utils";
import { ApiResponse, Product, ProductPayload } from "@/types";

const ensureData = <T>(resp: ApiResponse<T | null>): T => {
  if (resp.data == null) {
    throw new Error(resp.message || "服务器未返回数据");
  }
  return resp.data;
};

export async function searchProducts(keyword: string): Promise<Product[]> {
  try {
    const trimmed = keyword.trim();
    const resp = await invoke<ApiResponse<Product[] | null>>(
      "search_products",
      {
        keyword: trimmed.length > 0 ? trimmed : null,
      }
    );
    return resp.data ?? [];
  } catch (error) {
    throw new Error(getErrorMessage(error));
  }
}

export async function getProduct(productId: number): Promise<Product> {
  try {
    const resp = await invoke<ApiResponse<Product | null>>("get_product", {
      productId,
    });
    return ensureData(resp);
  } catch (error) {
    throw new Error(getErrorMessage(error));
  }
}

export async function createProduct(payload: ProductPayload): Promise<Product> {
  try {
    const resp = await invoke<ApiResponse<Product | null>>("create_product", {
      payload,
    });
    return ensureData(resp);
  } catch (error) {
    throw new Error(getErrorMessage(error));
  }
}

export async function updateProduct(
  productId: number,
  payload: ProductPayload
): Promise<Product> {
  try {
    const resp = await invoke<ApiResponse<Product | null>>("update_product", {
      productId: productId,
      payload,
    });
    return ensureData(resp);
  } catch (error) {
    throw new Error(getErrorMessage(error));
  }
}

export async function deleteProduct(productId: number): Promise<void> {
  try {
    const resp = await invoke<ApiResponse<null>>("delete_product", {
      // include both keys to match Rust command arg naming (some builds expect camelCase)
      productId: productId,
    });
    if (!resp.success) {
      throw new Error(resp.message || "删除商品失败");
    }
  } catch (error) {
    throw new Error(getErrorMessage(error));
  }
}

File: src-tauri/src/api/product.rs
use crate::{
    error::AppError,
    model::{ApiResponse, Product, ProductPayload},
    APP,
};

#[tauri::command]
pub async fn get_product(product_id: u32) -> Result<ApiResponse<Product>, AppError> {
    APP.get_product(product_id).await
}

#[tauri::command]
pub async fn search_products(
    keyword: Option<String>,
) -> Result<ApiResponse<Vec<Product>>, AppError> {
    APP.search_products(keyword).await
}

#[tauri::command]
pub async fn create_product(payload: ProductPayload) -> Result<ApiResponse<Product>, AppError> {
    APP.create_product(payload).await
}

#[tauri::command]
pub async fn update_product(
    product_id: u32,
    payload: ProductPayload,
) -> Result<ApiResponse<Product>, AppError> {
    APP.update_product(product_id, payload).await
}

#[tauri::command]
pub async fn delete_product(product_id: u32) -> Result<ApiResponse<()>, AppError> {
    APP.delete_product(product_id).await
}
